### PR TITLE
Upgrade MicroPython

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**
I've skipped to latest master as it has changes that require source
level updates here that I expect to be in the 2.1 release.

- Add new sound and power source files
- Update conversions to use new HAL constants
- Update sound expression HAL for function name change
- Stub the power HAL for now (all void returns)
- Update config to add power module

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -91,6 +94,7 @@ SRC_C += $(addprefix $(CODAL_PORT)/, \
 	modmusic.c \
 	modmusictunes.c \
 	modos.c \
+	modpower.c \
```
Comment (microbit-matt-hillsdon) [inline src/Makefile:95]: There's a bunch of new HAL for power that needs stubbing out.

Comment (microbit-matt-hillsdon) [reply src/Makefile:95]: e47392f3e9bff27ba52c2ff120deff0b93d6fead

Comment (microbit-matt-hillsdon) [inline src/Makefile:95]:
```python
>>> import power                                                                
Traceback (most recent call last):                                              
  File "<stdin>", line 1, in <module>                                           
ImportError: no module named 'power'       
```

...so I need to do something additional.

Comment (microbit-matt-hillsdon) [reply src/Makefile:95]: Confirmed it works in the beta.

Comment (microbit-matt-hillsdon) [reply src/Makefile:95]: Confirmed I have SoundEffect.copy so I haven't somehow got the old code.

Comment (microbit-matt-hillsdon) [reply src/Makefile:95]: 38564b5dd7212d73966d1b5595fa3726dec6f5e3

**Comment by microbit-matt-hillsdon** [comment]
We might need to port to the JS filesystem the behaviour change in main.c that clears the log whenever main.py is opened for writing.

We already clear the log on flash though, so I think this would just catch self-rewriting programs which might not be worth the effort. See https://github.com/microbit-foundation/micropython-microbit-v2/commit/6ade36b3cb26f6b945606a61763422ed630155f5 for the change. I'll check in with Carlos on this.

**Comment by microbit-matt-hillsdon** [comment]
@microbit-robert, as part of review please take a look at the diff:

https://github.com/microbit-foundation/micropython-microbit-v2/compare/318595a5e545e0176e106ff0db8562ca84fd622e...804c5fee4fd12e786e3815ef37aaefd5649153fa

..and see if there's anything I've missed that requires an update here.

File: src/Makefile [hunk 2]
```diff
@@ -59,6 +59,8 @@ SRC_C += \
 	main.c \
 	mphalport.c \
 
+# Upgrading?
+# Compare to lib/micropython-microbit-v2/src/codal_port/Makefile
```
Comment (microbit-matt-hillsdon) [inline src/Makefile]: I might move this to README.md as it could do with mentioning the other semi-duplicated files.

File: src/microbithal_js.c [hunk 3]
```diff
@@ -431,8 +451,8 @@ bool microbit_hal_audio_is_expression_active(void) {
     return mp_js_hal_audio_is_expression_active();
 }
 
-void microbit_hal_audio_play_expression_by_name(const char *name) {
-    mp_js_hal_audio_play_expression_by_name(name);
+void microbit_hal_audio_play_expression(const char *name) {
```
Comment (microbit-robert) [inline src/microbithal_js.c]: `*name` was also renamed to `*expr`.

Comment (microbit-matt-hillsdon) [reply src/microbithal_js.c]: Aligned in the commit 70c57d7c6b5ad733574beffdb8b71a21e1e168d3. Shouldn't have had any functional difference but makes sense to reduce the diff.

**Review by microbit-robert — APPROVED** [review]
LGTM. Filesystem change already noted in comments.

**Comment by microbit-matt-hillsdon** [comment]
Raised https://github.com/microbit-foundation/micropython-microbit-v2-simulator/issues/48 for the data log issue.

**Pull request merged**